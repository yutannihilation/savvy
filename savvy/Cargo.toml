[package]
name = "savvy"
description = "An unfriendly R extension interface"
version.workspace = true
edition.workspace = true
authors.workspace = true
license.workspace = true
repository.workspace = true
readme = "README.md"

[dependencies]
rlang-ffi-lite = { verion = "*", path = "../rlang-ffi-lite" }
once_cell = "1"

savvy-macro = { version = "0", path = "../savvy-macro" }

[build-dependencies]
cc = "1"

[package.metadata.dist]
dist = false
