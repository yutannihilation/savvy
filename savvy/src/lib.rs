//! # Savvy - An Unfriendly R Interface
//!

#![doc = include_str!("../docs/design.md")]

pub mod error;
pub mod protect;
pub mod sexp;
pub mod unwind_protect;

pub use error::{Error, Result};
pub use sexp::integer::{IntegerSxp, OwnedIntegerSxp};
pub use sexp::list::{ListElement, ListSxp, OwnedListSxp};
pub use sexp::logical::{LogicalSxp, OwnedLogicalSxp};
pub use sexp::null::NullSxp;
pub use sexp::real::{OwnedRealSxp, RealSxp};
pub use sexp::string::{OwnedStringSxp, StringSxp};
pub use sexp::Sxp;

pub use sexp::external_pointer::{get_external_pointer_addr, IntoExtPtrSxp};

pub use unwind_protect::unwind_protect;

// re-export
pub use rlang_ffi_lite::SEXP;
pub use savvy_macro::savvy;

use rlang_ffi_lite::{cetype_t_CE_UTF8, REprintf, Rf_allocVector, Rf_mkCharLenCE, Rprintf};

use std::ffi::CString;

// TODO: make this r_println! macro
pub fn r_print(msg: &str) -> crate::error::Result<SEXP> {
    unsafe {
        let msg_c_string = CString::new(msg).unwrap();
        unwind_protect(|| {
            Rprintf(msg_c_string.as_ptr());
            NullSxp.into()
        })
    }
}

pub fn r_eprint(msg: &str) -> crate::error::Result<SEXP> {
    unsafe {
        let msg_c_string = CString::new(msg).unwrap();
        unwind_protect(|| {
            REprintf(msg_c_string.as_ptr());
            NullSxp.into()
        })
    }
}

fn alloc_vector(arg1: u32, arg2: isize) -> crate::error::Result<SEXP> {
    unsafe { unwind_protect(|| Rf_allocVector(arg1, arg2)) }
}

// This wrapper function handles Error and panicks, and flag it by setting the
// lowest bit to 1. The lowest bit is supposed to be detected (and then removed)
// on the corresponding C function.
//
// cf. https://en.wikipedia.org/wiki/Tagged_pointer
pub fn handle_error(e: crate::error::Error) -> SEXP {
    match e {
        // The token is already tagged, so pass it as it is.
        error::Error::Aborted(token) => token,

        // In other cases, return the error string with the tag
        e => unsafe {
            let msg = e.to_string();
            let r_error = Rf_mkCharLenCE(
                msg.as_ptr() as *const i8,
                msg.len() as i32,
                cetype_t_CE_UTF8,
            );

            // set the error flag
            (r_error as usize | 1) as SEXP
        },
    }
}
