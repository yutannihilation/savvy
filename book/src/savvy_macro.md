# `#[savvy]` macro

This is a simple Rust function to add the specified suffix to the input
character vector. `#[savvy]` macro turns this into an R function.

```rust
use savvy::NotAvailableValue;   // for is_na() and na()

/// Add Suffix
/// 
/// @export
#[savvy]
fn add_suffix(x: StringSexp, y: &str) -> savvy::Result<savvy::Sexp> {
    let mut out = OwnedStringSexp::new(x.len())?;

    for (i, e) in x.iter().enumerate() {
        if e.is_na() {
            out.set_na(i)?;
            continue;
        }

        out.set_elt(i, &format!("{e}_{y}"))?;
    }

    out.into()
}
```

## Convention for a `#[savvy]` function

The example function above has this signature.

```rust
fn add_suffix(x: StringSexp, y: &str) -> savvy::Result<savvy::Sexp>
```

As you can guess, `#[savvy]` macro cannot be applied to arbitrary functions. The
function must satisfy the following conditions:

* The function's inputs can be
    * a non-owned savvy type (e.g., `IntegerSexp` and `RealSexp`)
    * a corresponding Rust type for scalar (e.g., `i32` and `f64`)
    * a user-defined struct marked with `#[savvy]` (`&T`, `&mut T`, or `T`)
    * a user-defined enum marked with `#[savvy]` (`&T`, or `T`)
* The function's return value must be either
    * `savvy::Result<()>` for the case of no actual return value
    * `savvy::Result<savvy::Sexp>` for the case of some return value of R object
    * `savvy::Result<T>` for the case of some return value of a user-defined
      struct or enum marked with `#[savvy]`

## How things work under the hood

If you mark a funtion with `#[savvy]` macro, the corresponding implementations are generated:

1. Rust functions
    1. a wrapper function to handle Rust and R errors gracefully
    2. a function with the original body and some conversion from raw `SEXP`s to savvy types.
2. C function signature for the Rust function
3. C implementation for bridging between R and Rust
4. R implementation

For example, the above implementation generates the following codes. (`#[savvy]`
macro can also be used on `struct` and `enum`, but let's focus on function's
case for now for simplicity.)

### Rust functions

(The actual code is a bit more complex to handle possible `panic!` properly.)

```rust
#[allow(clippy::missing_safety_doc)]
#[no_mangle]
pub unsafe extern "C" fn add_suffix(x: SEXP, y: SEXP) -> SEXP {
    match savvy_add_suffix_inner(x, y) {
        Ok(result) => result.0,
        Err(e) => savvy::handle_error(e),
    }
}

unsafe fn savvy_add_suffix_inner(x: SEXP, y: SEXP) -> savvy::Result<savvy::Sexp> {
    let x = <savvy::RealSexp>::try_from(savvy::Sexp(x))?;
    let y = <&str>::try_from(savvy::Sexp(y))?;
    
    // ...original body...

}
```

### C function signature

```c
SEXP add_suffix(SEXP x, SEXP y);
```

### C implementation

(let's skip the details about `handle_result` for now)

```c
SEXP add_suffix__impl(SEXP x, SEXP y) {
    SEXP res = add_suffix(x, y);
    return handle_result(res);
}
```

### R implementation

The Rust comments with three slashes (`///`) is converted into Roxygen comments
on R code.

```r
#' Add Suffix
#' 
#' @export
add_suffix <- function(x, y) {
  .Call(add_suffix__impl, x, y)
}
```

## Using `#[savvy]` on other files than `lib.rs`

You can use `#[savvy]` macro just the same as `lib.rs`. One thing you have to do
is that you need to export the objects by `*`. This is because `#[savvy]`
defines additional functions other than the original ones, and some of these
also need to be exported. Since you don't know the names of such auto-generated
functions, `*` is the solution.

```rust
mod foo;
pub use foo::*;
```
