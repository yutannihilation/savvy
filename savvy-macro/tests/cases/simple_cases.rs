#![allow(unused_variables)]

use savvy_macro::savvy;

#[savvy]
fn no_return_type(x: i32) {}

#[savvy]
fn wrong_return_type(x: i32) -> i32 {}

#[savvy]
fn wrong_type_owned_int(x: OwnedIntegerSexp) -> savvy::Result<()> {
    Ok(())
}

#[savvy]
fn wrong_type_owned_real(x: OwnedRealSexp) -> savvy::Result<()> {
    Ok(())
}

#[savvy]
fn wrong_type_owned_logical(x: OwnedLogicalSexp) -> savvy::Result<()> {
    Ok(())
}

#[savvy]
fn wrong_type_owned_string(x: OwnedStringSexp) -> savvy::Result<()> {
    Ok(())
}

struct Foo1;

#[savvy]
impl Foo1 {
    fn foo_self(&self) -> Self {
        Self {}
    }
}

struct Foo2;

#[savvy]
impl Foo2 {}

#[savvy]
fn wrong_type_custom_type_no_ref(x: Foo2) -> savvy::Result<()> {
    Ok(())
}

#[savvy]
fn correct_type_custom_type_ref(x: &Foo2) -> savvy::Result<()> {
    Ok(())
}

#[savvy]
fn correct_type_custom_type_ref_mut(x: &mut Foo2) -> savvy::Result<()> {
    Ok(())
}

fn main() {}
