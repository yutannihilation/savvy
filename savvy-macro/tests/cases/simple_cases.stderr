error: function must have return type
 --> tests/cases/simple_cases.rs:5:1
  |
5 | #[savvy]
  | ^^^^^^^^
  |
  = note: this error originates in the attribute macro `savvy` (in Nightly builds, run with -Z macro-backtrace for more info)

error: the return type must be either syn::Result<Sexp> or syn::Result<()>
 --> tests/cases/simple_cases.rs:9:30
  |
9 | fn wrong_return_type(x: i32) -> i32 {}
  |                              ^^^^^^

error: `Owned-` types are not allowed here. Did you mean `IntegerSexp`?
  --> tests/cases/simple_cases.rs:12:28
   |
12 | fn wrong_type_owned_int(x: OwnedIntegerSexp) -> savvy::Result<()> {
   |                            ^^^^^^^^^^^^^^^^

error: `Owned-` types are not allowed here. Did you mean `RealSexp`?
  --> tests/cases/simple_cases.rs:17:29
   |
17 | fn wrong_type_owned_real(x: OwnedRealSexp) -> savvy::Result<()> {
   |                             ^^^^^^^^^^^^^

error: `Owned-` types are not allowed here. Did you mean `LogicalSexp`?
  --> tests/cases/simple_cases.rs:22:32
   |
22 | fn wrong_type_owned_logical(x: OwnedLogicalSexp) -> savvy::Result<()> {
   |                                ^^^^^^^^^^^^^^^^

error: `Owned-` types are not allowed here. Did you mean `StringSexp`?
  --> tests/cases/simple_cases.rs:27:31
   |
27 | fn wrong_type_owned_string(x: OwnedStringSexp) -> savvy::Result<()> {
   |                               ^^^^^^^^^^^^^^^

error: For safety, a function that takes `self` and returns `Self` is not allowed
  --> tests/cases/simple_cases.rs:35:5
   |
35 | /     fn foo_self(&self) -> Self {
36 | |         Self {}
37 | |     }
   | |_____^

error: A user-defined struct must be in the form of either `&Foo2` or `&mut Foo2`
  --> tests/cases/simple_cases.rs:46:37
   |
46 | fn wrong_type_custom_type_no_ref(x: Foo2) -> savvy::Result<()> {
   |                                     ^^^^
